flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/EpochManager.java

```diff
@@ -0,0 +1,207 @@
+         */
+        PARALLEL_BETWEEN_EPOCH
+    }
+
+    /**
+     * The reference to the {@link AsyncExecutionController}, used for {@link
+     * ParallelMode#SERIAL_BETWEEN_EPOCH}. Can be null when testing.
+     */
+    final AsyncExecutionController<?> asyncExecutionController;
+
+    /** The number of epochs that have arrived. */
+    long epochNum;
+
+    /** The output queue to hold ongoing epochs. */
+    LinkedList<Epoch> outputQueue;
+
+    /** Current active epoch, only one active epoch at the same time. */
+    Epoch activeEpoch;
+
+    public EpochManager(AsyncExecutionController<?> aec) {
+        this.epochNum = 0;
+        this.outputQueue = new LinkedList<>();
+        this.asyncExecutionController = aec;
+        // init an empty epoch, the epoch action will be updated when non-record is received.
+        initNewActiveEpoch();
+    }
+
+    /**
+     * Add a record to the current epoch and return the current open epoch, the epoch will be
+     * associated with the {@link RecordContext} of this record. Must be invoked within task thread.
+     *
+     * @return the current open epoch.
+     */
+    public Epoch onRecord() {
+        activeEpoch.ongoingRecordCount++;
+        return activeEpoch;
+    }
+
+    /**
+     * Add a non-record to the current epoch, close current epoch and open a new epoch. Must be
+     * invoked within task thread.
+     *
+     * @param action the action associated with this non-record.
+     * @param parallelMode the parallel mode for this epoch.
+     */
+    public void onNonRecord(Runnable action, ParallelMode parallelMode) {
+        if (parallelMode == ParallelMode.SERIAL_BETWEEN_EPOCH) {
+            asyncExecutionController.drainInflightRecords(0);
+        }
+        activeEpoch.close(action);
+        if (!outputQueue.isEmpty() || !activeEpoch.tryFinish()) { // which means the first epoch
+            outputQueue.offer(activeEpoch);
+        }
+        LOG.trace(
+                "on NonRecord, old epoch: {}, outputQueue size: {}",
+                activeEpoch,
+                outputQueue.size());
+        initNewActiveEpoch();
+    }
+
+    /**
+     * Complete one record in the specific epoch. Must be invoked within task thread.
+     *
+     * @param epoch the specific epoch
+     */
+    public void completeOneRecord(Epoch epoch) {
+        if (--epoch.ongoingRecordCount == 0) {
+            // If one epoch has been closed before and all records in
+            // this epoch have finished, the epoch will be removed from the output queue.
+            while (!outputQueue.isEmpty() && outputQueue.peek().tryFinish()) {
+                LOG.trace(
+                        "Finish epoch: {}, outputQueue size: {}",
+                        outputQueue.peek(),
+                        outputQueue.size());
+                outputQueue.pop();
+            }
+        }
+    }
+
+    private void initNewActiveEpoch() {
+        this.activeEpoch = new Epoch(epochNum++);
```

Comment:
I'd suggest a tool method `tryFinishInQueue` which does:
```
while (!outputQueue.isEmpty() && outputQueue.peek().tryFinish()) {
                LOG.trace(
                        "Finish epoch: {}, outputQueue size: {}",
                        outputQueue.peek(),
                        outputQueue.size());
                outputQueue.pop();
            }
```

and in `initNewActiveEpoch`, it does:
```
outputQueue.offer(activeEpoch);
activeEpoch = new Epoch(epochNum++);
tryFinishInQueue();
```